bin/fin

```diff
@@ -6261,6 +6343,630 @@ set_project_root_permissions ()
 	is_root && chown -R 1000:1000 "$PROJECT_ROOT"
 }
 
+# Pull main command for interacting with Hosting Providers.
+provider_pull ()
+{
+	eval $(parse_params "$@")
+	if [[ "${HOSTING_PLATFORM}" == "" || "${HOSTING_SITE}" == ""  ]]; then
+		echo-error "Hosting Provider not set up. Following variables must be set: HOSTING_PLATFORM and HOSTING_SITE"
+		exit 1
+	fi
+
+	HOSTING_PLATFORM=$(echo ${HOSTING_PLATFORM} | tr '[:upper:]' '[:lower:]')
+
+	# Check to see if a support platform is accepted.
+	if [[ "${HOSTING_PLATFORM}" != "acquia" ]] && [[ "${HOSTING_PLATFORM}" != "pantheon" ]] && [[ "${HOSTING_PLATFORM}" != "platformsh" ]] &&
+		[[ "${HOSTING_PLATFORM}" != "drush" ]] && [[ "${HOSTING_PLATFORM}" != "wp" ]]; then
+		echo-error "Only the following platforms are supported: Acquia (acquia), Pantheon (pantheon), Platform.sh (platformsh), Drupal (drush), and WordPress (wp)."
+		exit 1
+	fi
+
+	# Run some checks before continuing.
+	provider_pull_find_function ${HOSTING_PLATFORM} 'check'
+
+	echo-green "Starting provider pull for ${HOSTING_PLATFORM}"
+
+	# Query the provider for the given repo url. Only Pantheon, Platform.sh, and Acquia support this.
+	# All others will need to do a git clone.
+	if [[ "${ARGV[0]}" == "init" ]]; then
+		provider_pull_init
+		return
+	fi
+
+	# Check to see if providers require anything else.
+	provider_pull_find_function ${HOSTING_PLATFORM} 'precheck'
```

Comment:
1. Why is there 2 "check requirements" functions?
2. Precheck function in all provides seem to always use generic precheck
3. Why is precheck executed after check, which is semantically weird, since it's precheck should be the one executed first.